### PR TITLE
Workspace resource

### PR DESCRIPTION
**Pull request by ilangofman**
Issue #, if available: [1323](https://github.com/aws-controllers-k8s/community/issues/1323)

Description of changes:

Added the workspace resource.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/workspace/hooks.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	"context"
+	"errors"
+
+	"github.com/aws-controllers-k8s/prometheusservice-controller/apis/v1alpha1"
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackcondition "github.com/aws-controllers-k8s/runtime/pkg/condition"
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	svcsdk "github.com/aws/aws-sdk-go/service/prometheusservice"
+	corev1 "k8s.io/api/core/v1"
```
Comment (a-hilaly) [inline pkg/resource/workspace/hooks.go]: Can you please split the local import paths from the external ones? we try to follow the standards mentioned in https://github.com/aws-controllers-k8s/lambda-controller/blob/main/pkg/resource/function/hooks.go#L117

File: pkg/resource/workspace/hooks.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	exit := rlog.Trace("rm.updateWorkspaceTags")
+	defer exit(err)
+
+	added, removed, updated := compareMaps(latest.ko.Spec.Tags, desired.ko.Spec.Tags)
+
+	if len(removed) > 0 {
+		removeTags := []*string{}
+		for i := range removed {
+			removeTags = append(removeTags, &removed[i])
+		}
```
Comment (a-hilaly) [inline pkg/resource/workspace/hooks.go:114]: Great catch on this one!

File: pkg/resource/workspace/hooks.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+	// Check if the state is active before updating
+	if !workspaceActive(latest) {
+		msg := "Cannot update workspace as it is not active, current status=" + string(*latest.ko.Status.Status.StatusCode)
+		ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
+		return desired, ackrequeue.NeededAfter(
+			errors.New(msg),
+			ackrequeue.DefaultRequeueAfterDuration,
+		)
+	}
```
Comment (a-hilaly) [inline pkg/resource/workspace/hooks.go:77]: 👍

File: go.mod [hunk 4]
```diff
@@ -1,10 +1,13 @@
 module github.com/aws-controllers-k8s/prometheusservice-controller
 
 go 1.17
+
 require (
 	github.com/aws-controllers-k8s/runtime v0.18.4
```
Comment (a-hilaly) [inline go.mod]: We just released v0.19.3. Can you please send a PR to the amp controller repository making just a runtime update?

File: test/e2e/__init__.py [hunk 5]
```diff
@@ -18,15 +18,15 @@
 from acktest.resources import load_resource_file
 
 SERVICE_NAME = "Prometheus"
-CRD_GROUP = "prometheus.services.k8s.aws"
+CRD_GROUP = "prometheusservice.services.k8s.aws"
```
Comment (a-hilaly) [inline test/e2e/__init__.py:21]: ++

File: test/e2e/condition.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+import pytest
+
+from acktest.k8s import resource as k8s
+
+CONDITION_TYPE_ADOPTED = "ACK.Adopted"
+CONDITION_TYPE_RESOURCE_SYNCED = "ACK.ResourceSynced"
+CONDITION_TYPE_TERMINAL = "ACK.Terminal"
+CONDITION_TYPE_RECOVERABLE = "ACK.Recoverable"
+CONDITION_TYPE_ADVISORY = "ACK.Advisory"
+CONDITION_TYPE_LATE_INITIALIZED = "ACK.LateInitialized"
```
Comment (a-hilaly) [inline test/e2e/condition.py]: This tools have been moved to https://github.com/aws-controllers-k8s/test-infra/blob/main/src/acktest/k8s/condition.py you also just bump the testing framework version and start using it :)

File: test/e2e/tests/test_workspace.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+from e2e import service_marker, CRD_GROUP, CRD_VERSION, load_prometheusservice_resource
+from e2e.replacement_values import REPLACEMENT_VALUES
+from e2e.bootstrap_resources import get_bootstrap_resources
+from e2e import condition
+
+
+RESOURCE_KIND = "Workspace"
+RESOURCE_PLURAL = "workspaces"
+
+CREATE_WAIT_AFTER_SECONDS = 30
```
Comment (a-hilaly) [inline test/e2e/tests/test_workspace.py]: Surprising that workspaces can take this much to get created..

Comment (a-hilaly) [reply test/e2e/tests/test_workspace.py]: I just realized that these values are not used, you can removed them

**Review by a-hilaly — COMMENTED** [review]
Excellent work on this, Ilan!
Left few nit comments inline

**Comment by ilangofman** [comment]
Merged on the upstream branch.

**Pull request closed without merging**